test/e2e/pkg/infra/infra.go

```diff
@@ -0,0 +1,84 @@
+package infra
+
+import (
+	"context"
+
+	e2e "github.com/tendermint/tendermint/test/e2e/pkg"
+)
+
+// TestnetInfra provides an API for manipulating the infrastructure of a
+// specific testnet.
+type TestnetInfra interface {
+	//
+	// Overarching testnet infrastructure management.
+	//
+
+	// Setup generates any necessary configuration for the infrastructure
+	// provider during testnet setup.
+	Setup(ctx context.Context) error
+
+	// Stop will stop all running processes throughout the testnet without
+	// destroying any infrastructure.
+	Stop(ctx context.Context) error
+
+	// Pause will pause all processes in the testnet.
+	Pause(ctx context.Context) error
+
+	// Unpause will resume a paused testnet.
+	Unpause(ctx context.Context) error
+
+	// ShowLogs prints all logs for the whole testnet to stdout.
+	ShowLogs(ctx context.Context) error
+
+	// TailLogs tails the logs for all nodes in the testnet, if this is
+	// supported by the infrastructure provider.
+	TailLogs(ctx context.Context) error
+
+	// Cleanup stops and destroys all running testnet infrastructure and
+	// deletes any generated files.
+	Cleanup(ctx context.Context) error
+
+	//
+	// Node management, including node infrastructure.
+	//
+
+	// StartNode provisions infrastructure for the given node and starts it.
+	StartNode(ctx context.Context, node *e2e.Node) error
```

Comment:
So, in the case of Digital Ocean, provisioning infrastructure is one of the slow steps and I think that doing it all at once with a single or small set of calls to DO is likely going to be pretty advantageous. I think we should add some mechanism to the API to do the provisioning all at once. Perhaps that can happen in `Setup`, that seems fine to me, but I just wanted to call it out. Perhaps this method will start the node, provisioning any infrastructure if necessary?

Reply:
This is likely to be the case on any VM-based provider. 

I'm also having sort of flashbacks to all of the ways this can go wrong and get aborted or end up leaking nodes at great expense. 

In addition to figuring out how this is going to work for DO (presuming that we're just calling out to terraform, perhaps its better to think of this as a "terraform provider" rather than a DO provider,) perhaps it makes sense to make sure that our interface design decisions would make sense for a k8s based deploy (just as an exercise.)

Reply:
> I think we should add some mechanism to the API to do the provisioning all at once.

The only real way to accomplish this is to rearchitect the testnet runner. I initially started doing this over the weekend but thought everyone may be more amenable to reviewing a smaller change, but I'd be happy to revive my alternative approach as I still have it stashed locally. It's just going to be a bigger PR and will look very different to the current one :slightly_smiling_face: 

But that approach would involve introducing a higher-level interface:

```golang
type Runner interface {
    Start(ctx context.Context) error
    Setup(ctx context.Context) error
    Cleanup(ctx context.Context) error
    // etc. - only testnet-level manipulation and not node- or process-level
}
```

> I'm also having sort of flashbacks to all of the ways this can go wrong and get aborted or end up leaking nodes at great expense.

I don't think we need to be too concerned about this. It's easy to delete a DO "project" and all of its associated resources.

> In addition to figuring out how this is going to work for DO (presuming that we're just calling out to terraform, perhaps its better to think of this as a "terraform provider" rather than a DO provider,) perhaps it makes sense to make sure that our interface design decisions would make sense for a k8s based deploy (just as an exercise.)

Having built and operated several k8s clusters the hard way from scratch, I'd be very much against using k8s for any of this testnet infra. I'd say we'd need a _really_ good justification to even think of supporting k8s, and I don't see one at present that provides any advantage over a DO or AWS-based approach (a DO-based approach can be far more easily translated to AWS).